File: src/main/java/algo/breadthFirstSearch/TimeNeededToInformAllEmployees_1376.java
package algo.breadthFirstSearch;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class TimeNeededToInformAllEmployees_1376 {
  public static int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
    var adjMap = new HashMap<Integer, List<Integer>>();
    for (var i = 0; i < manager.length; i++) {
      if (!adjMap.containsKey(manager[i])) {
        adjMap.put(manager[i], new ArrayList<>());
      }
      adjMap.get(manager[i]).add(i);
    }

    var time = 0;
    var q = new ArrayDeque<EmployeeTime>();
    q.offer(new EmployeeTime(headID, 0));
    while (!q.isEmpty()) {
      var mng = q.poll();
      time = Math.max(time, mng.time());
      if (adjMap.containsKey(mng.employeeId())) {
        for (var emp : adjMap.get(mng.employeeId())) {
          q.offer(new EmployeeTime(emp, mng.time() + informTime[mng.employeeId()]));
        }
      }
    }
    return time;
  }

  private record EmployeeTime (int employeeId, int time) {}

  public static void main(String[] args) {
    System.out.println(numOfMinutes(6, 2, new int[]{2,2,-1,2,2,2}, new int[]{0,0,1,0,0,0}));
  }
}

File: src/main/kotlin/algo/breadthFirstSearch/TimeNeededToInformAllEmployees_1376.kt
package algo.breadthFirstSearch

import java.util.LinkedList

class KtTimeNeededToInformAllEmployees_1376 {
    fun numOfMinutes(n: Int, headID: Int, manager: IntArray, informTime: IntArray): Int {
        val adjMap: MutableMap<Int, MutableList<IntArray>> = HashMap()
        manager.indices.forEach {
            adjMap.getOrPut(manager[it]) { mutableListOf() }.add(intArrayOf(it, informTime[it]))
        }

        var time = 0
        val q = LinkedList<IntArray>()
        q.offer(intArrayOf(headID, informTime[headID]))
        while (q.isNotEmpty()) {
            val size = q.size
            for (i in 1..size) {
                val emp = q.poll()
                val (empId, iTime) = emp
                time = time.coerceAtLeast(iTime)
                adjMap[empId]?.forEach { q.offer(intArrayOf(it[0], it[1] + iTime)) }
            }
        }
        return time
    }
}

fun main() {
    val inform = KtTimeNeededToInformAllEmployees_1376()
    println(inform.numOfMinutes(6, 2, intArrayOf(2,2,-1,2,2,2), intArrayOf(0,0,1,0,0,0)))
}